text/0000-let-expression.md

```diff
@@ -0,0 +1,807 @@
+- Feature Name: let_expression
+- Start Date: 2021-08-08
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Convert `let <pat> = <expr>` from a statement to an expression of type `bool`. Make
+the language more consistent and remove corners by generalizing, plus adding
+a bunch of magic constructs that are useful. After this RFC, you'll be able to write, among many other things:
+
+```rust
+// generalized matches! macro:
+assert!((let Some(x) = y) && x > 2);
+
+// generalized let-else construct
+(let Some(a) = b) && (let Some(c) = f(&a)) || {
+    return Err("failed");
+};
+println!("{}, {}", a, c);
+
+// generalized assignment with default
+(let Some(x) = y) || (let Foo(x) = bar) || (let x = default);
+println!("{}", x);
+```
+
+# Motivation
+[motivation]: #motivation
+
+The main motivation for this RFC is improving consistency and ergonomics of language.
+
+## Consistency
+
+Currently we have `if`, `if let` and `if let && let` and we teach them as three different
+constructs (plus `let else` in future). But if we make `let` a `bool` expression, all will become the same and it would be
+easier for new learners to get it. After this RFC you get only an unused parenthesis warning
+for `if (let Some(x) = y) { }`, not a hard error. And it will have the same behavior with
```

Comment:
I think that it's worth adding a helpful error message for this even if the RFC is not accepted. We could even consider making the compiler recover, but with a warning, though I'm not sure that's the way to go. I haven't checked what error message is given currently though.

Reply:
I think recovering with a warning is good, especially in if-let-chains because people may want to sure that `&&` isn't interpreted wrong so add a (unnecessary) parenthesis. But it will make demand for let expression even more, because users will say: What is this `let` that I can put it in `if` with parenthesis and `&&` it inside if, but nothing else?